truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/impl/DefaultTruffleRuntime.java

```diff
@@ -156,12 +158,30 @@ public FrameInstance getCurrentFrame() {
         return result;
     }
 
-    void pushFrame(FrameInstance frame) {
-        getThreadLocalStackTrace().addFirst(frame);
+    void pushFrame(VirtualFrame frame, CallTarget target) {
+        LinkedList<FrameInstance> threadLocalStackTrace = getThreadLocalStackTrace();
+        threadLocalStackTrace.addFirst(new DefaultFrameInstance(frame, target, null));
+    }
+
+    void pushFrame(VirtualFrame frame, CallTarget target, Node parentCallNode) {
+        LinkedList<FrameInstance> threadLocalStackTrace = getThreadLocalStackTrace();
+        // we need to ensure that frame instances are immutable to we need to recreate the parent
+        // frame
+        if (threadLocalStackTrace.size() > 0) {
+            DefaultFrameInstance oldInstance = (DefaultFrameInstance) threadLocalStackTrace.removeFirst();
+            threadLocalStackTrace.addFirst(new DefaultFrameInstance(oldInstance.getFrame(), oldInstance.getCallTarget(), parentCallNode));
```

Comment:
This part is a bit messy, part of why I prefer getCallNode() to be the call site, but compatibility is more important.